src/Analyzer/QueryTreeBuilder.cpp

```diff
@@ -233,22 +233,43 @@ QueryTreeNodePtr QueryTreeBuilder::buildSelectExpression(const ASTPtr & select_q
     auto select_settings = select_query_typed.settings();
     SettingsChanges settings_changes;
 
-    if (is_subquery)
+    /// We are going to remove settings LIMIT and OFFSET and
+    /// further replace them with corresponding expression nodes
+    UInt64 limit = 0;
+    UInt64 offset = 0;
+
+    /// remove global settings limit and offset
+    if (const Settings & settings_ref = updated_context->getSettingsRef(); settings_ref.limit || settings_ref.offset)
```

Comment:
Consider to rename `settings_ref`, into just `const auto & settings =`.

Reply:
it's only used to check if limit or offset present and further we use `settings` name for copy